src/main/kotlin/org/rust/lang/core/resolve/NameResolution.kt

```diff
@@ -362,8 +362,32 @@ fun processPathResolveVariants(lookup: ImplLookup, path: RsPath, isCompletion: B
             if (processor("self", base)) return true
         }
         if (processItemOrEnumVariantDeclarations(base, ns, processor, withPrivateImports = isSuperChain)) return true
-        if (base is RsTypeDeclarationElement && parent !is RsUseSpeck) {
-            // Foo::<Bar>::baz
+        if (base is RsTypeDeclarationElement && parent !is RsUseSpeck) { // Foo::<Bar>::baz
+            val shadowingProcessor = run {
+                // Self-qualified type paths `Self::Item` inside impl items are restricted to resolve
+                // to only members of the current impl or implemented trait or its parent traits
+                if (Namespace.Types in ns && base is RsImplItem && qualifier.hasCself) {
+                    NameResolutionTestmarks.selfRelatedTypeSpecialCase.hit()
+                    val traits = base.implementedTrait?.flattenHierarchy
+                        ?.map { it.foldTyTypeParameterWith { TyInfer.TyVar(it) } }
+                        ?: return@run { e: AssocItemScopeEntry -> processor(e) }
+
+                    fun(e: AssocItemScopeEntry): Boolean {
+                        if (e.element !is RsTypeAlias) return processor(e)
+
+                        val implementedTrait = e.source.value.implementedTrait
+                            ?.foldTyTypeParameterWith { TyInfer.TyVar(it) }
+                            ?: return processor(e)
+
+                        val isAppropriateTrait = traits.any {
+                            lookup.ctx.probe { lookup.ctx.combineBoundElements(it, implementedTrait) }
+                        }
+                        return if (isAppropriateTrait) processor(e) else false
+                    }
+                } else {
+                    { e: AssocItemScopeEntry -> processor(e) }
```

Comment:
Looks like we can just return `processor` here (and above), because `(ScopeEntry) -> Boolean` is contravariant to `(AssocItemScopeEntry) -> Boolean`.

Reply:
I'd do that if it were typechecked :(